### DISCUSSION TITLE
Add canonical URL of each page

### DISCUSSION THREAD

**Discussion by anuj9196** [analogjs/analog#1311] [General]
### Which scope/s are relevant/related to the feature request?

astro-angular

### Information

After generating a static site, the URLs are visited using query parameters, which are used to track analytics. However, Google is reporting them as separate URLs.

Can a canonical URL element be added to the root URL during the static build? If not, a feature to do so would be of a great addition.

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [X] No

**Comment by ajitzero** [comment]
If this is an Astro app using `astro-angular` (as mentioned above), you can add the canonical URL with your custom logic in Astro in your base layout or header fragment.

If this is an Analog app (and if the above scope is incorrectly set to astro-angular), then this may be blocked due to an upstream issue: https://github.com/unjs/nitro/issues/2052

An unofficial solution would be to modify the document in the constructor, which would be pre-computed for SSG pages at build time. I'm not sure about the SSR pages.

**Comment by anuj9196** [comment]
Got it temporarily solved by adding to `postRenderer`
```typescript
// Base URL of your website
const baseUrl = 'https://tapbeep.com';
// Generate the canonical URL based on the route
const canonicalUrl = `${baseUrl}${route.route}`;
// Inject the canonical tag into the <head> section
const canonicalTag = `<link rel="canonical" href="${canonicalUrl}">`;
route.contents = route.contents?.replace('</head>', `${canonicalTag}</head>`);
```

**Comment by brandonroberts** [comment]
Cool, we'll note the workaround for now

**Comment by anuj9196** [comment]
Above shared solutions adds trailing slash to the base url `https://tapbeep.com/`.

Further modified it to remove trailing slash in case of root URL

```typescript
let _route = route.route;
if (_route === '/') {
    _route = '';
}
const canonicalUrl = `${baseUrl}${_route}`;        
```
